### PR TITLE
[6x] Project outrrefs in the targetlist of subqueries in ORCA

### PR DESCRIPTION
**Pull request by hardikar**
In a previous commit (df5f06d), we added code to fall back gracefully
when a subquery select list contains a single outer ref that is not part
of an expression, such as in

`select * from foo where a is null or a = (select foo.b from bar)`

This commit adds a fix that allows us to handle such queries in ORCA
by adding a project in the translator that will echo the outer ref
from within the subquery, and using that projected value in the
select list of the subquery. This ensures that we use a NULL value
for the outer ref when the subquery returns no rows.

Also note that this is still skipped for grouping cols in the target
list. This was done to avoid regression for certain queries, such as:

```
select *
from A
where not exists (select sum(C.i)
                  from C
                  where C.i = A.i
                  group by a.i);
```

ORCA is currently unable to decorrelate sub-queries that contain project
nodes, So, a `SELECT 1` in the subquery would also cause this
regression. In the above query, the parser adds `a.i` to the target list
of the subquery, that would get an echo projection (as described above),
and thus would prevent decorelation by ORCA. For this reason, we decided
to maintain existing behavior until ORCA is able to handle projections
in subqueries better.

### CODE REVIEW

**Review by zellerh — COMMENTED** [review]
LGTM, very nice fix that I believe is safe. But, given that I'm a co-author, I can't approve it.

**Comment by d** [comment]
Hi @hardikar what's the best time I can push a (rebased and) reformatted history to your PR so that it doesn't disrupt what you're working on?

**Comment by d** [comment]
Hi @hardikar I've pushed a (rebased and) reformatted version of this PR's history. For posterity I'm also hosting the two versions on my fork:

before: https://github.com/d/gpdb/tree/10863-backup
after: https://github.com/d/gpdb/tree/10863-fmt

File: src/test/regress/expected/gporca_optimizer.out [hunk 1]
```diff
@@ -12875,33 +12875,25 @@ WHERE  tcorr1.a IS NULL OR
                            FROM   tcorr2) al
                    WHERE  3 = tcorr1.a
                   );
-INFO:  GPORCA failed to produce a plan, falling back to planner
```
Comment (chrishajas) [inline src/test/regress/expected/gporca_optimizer.out:12878]: change all these comments since they no longer fall back.

**Review by chrishajas — APPROVED** [review]
LGTM. 

Is the previous fallback logic in df5f06d still relevant/possible to hit (possibly with a grouping col?)? If so do we have an ICG test for this? Will skip-level outer refs also be handled with this?

**Review by dgkimura — APPROVED** [review]
Looks like a neat fix.  And I like the detail you put in the commit message.  Though admittedly, I'm a little confused reading this part..

> This ensures that we use a NULL value for the outer ref when the subquery returns no rows.

Is it when subquery returns no rows or when subquery doesn't contain any "inner" ref columns?

**Comment by d** [comment]
> This ensures that we use a NULL value
> for the outer ref when the subquery returns no rows.

This took more than half a second for me to parse. ITYM:

This ensures that the subquery expression (scalar subquery in ORCA parlance) is of NULL value when the subquery returns 0 rows

File: src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp [hunk 2]
```diff
@@ -3866,8 +3866,18 @@ CTranslatorQueryToDXL::TranslateTargetListToDXLProject(
 			CTranslatorUtils::IsGroupingColumn(target_entry, plgrpcl);
 		if (!is_groupby || (is_groupby && is_grouping_col))
 		{
+			// Insist projection for any outer refs to ensure any decorelation of a
+			// subquery results in a correct plan using the projected reference,
+			// instead of the outer ref directly.
+			// TODO: Remove is_grouping_col from this check once const projections in
+			// subqueries no longer prevent decorrelation
```
Comment (d) [inline src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp:3873]: :fireworks: :+1: thanks for adding the comments. It always takes me a long time to reason why `is_grouping_col` was in the condition whenever I'm near this code

**Comment by d** [comment]
I'm reviewing this

File: src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp [hunk 3]
```diff
@@ -3866,8 +3866,18 @@ CTranslatorQueryToDXL::TranslateTargetListToDXLProject(
...
+			BOOL insist_proj =
+				(IsA(target_entry->expr, Var) &&
+				 ((Var *) (target_entry->expr))->varlevelsup > 0 &&
+				 !is_grouping_col);
 			CDXLNode *project_elem_dxlnode = TranslateExprToDXLProject(
-				target_entry->expr, target_entry->resname);
+				target_entry->expr, target_entry->resname,
+				insist_proj /* insist_new_colids */);
```
Comment (d) [inline src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp:3880]: Stupid question, what difference does it make if we do not allocate a new column reference and instead use the existing col id?

Comment (d) [reply src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp:3880]: Hmm, I can answer it myself: if we don't allocate a new col id, this won't have been a param, right?

Comment (hardikar) [reply src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp:3880]: Right. Also, we need to use the new colid after projecting, else (I believe) the projection remains unused.

**Comment by hardikar** [comment]
@dgkimura

File: src/test/regress/expected/gporca_optimizer.out (hunk 1 above)
Comment (hardikar) [reply src/test/regress/expected/gporca_optimizer.out:12878]: Done. Thanks.

**Comment by d** [comment]
IIUC, this PR fixes two things:

1. It stops falling back for a class of queries
1. It avoids an incorrect decorrelation by adding a layer of projection into the subquery.

I have no objection to shipping this for expediency.

But a superfluous project is supposed to be a no-op. It bothers me that none of us quite understand why the subquery handling would incorrectly decorrelate the subquery otherwise. We should follow up on that after shipping this.

**Pull request merged**